### DISCUSSION TITLE
Trouble with modelica model creation

### DISCUSSION THREAD

**Discussion by hamadah10** [OpenModelica/OpenModelica#10771] [Q&A]
Hello,

I am new to modelica and working on a project that requires a use a PID controller to control a car model created using modelica. 

With the help with some people, I managed to discover how rather good and working example to work with. Now, in order to achieve my goal, I wanted to modify the example further with two things. 

1- A PID controller to (replace) the engine or something in order to provide the necessary controls for the car to function.

2- This is a more advanced and complex application which is the addition a GPS system into the model. At this stage, I am only looking for a starting insight on how i might get started on this. 

Here is a picture of the model and what I want to do

![Screenshot 2023-05-30 110533](https://github.com/OpenModelica/OpenModelica/assets/135056629/e737e72a-0b26-4f01-a2ad-bbf5091db694)

the translational error that is on the screen is related to when attempted to use the pid controller instead of the engine directly. I know it is not how this works so it can simply be ignored here.

**Comment by casella** [comment]
@hamadah10 I don't get what you mean with "replace the engine with PID controller". An engine model has physical connectors, a PID controller has input/output physical connectors. These are incompatible, you can't connect them together. Also, it is not clear to me how an engine could be represented by a PID...

**Reply by hamadah10** [reply]
@casella Sorry I have not explained this correctly.

Here is what I am trying to do.

I have a very simply physical model for an autonomous vehicle.

The components I am trying to fit together in a model are as follows :

Sensors (to simply make all the components work/communicate together)

"Flight control" or state estimator

Pid controller

Motor then something to update the flight control with the new 'location' and go around the loop once again.

Its my first project on a car ever and on modelica as well which Is why I'm making all sort of errors here.

I'm not sure if I would need a motor in order to make a pid controller work in such setting. The reason I'm using pid controller in the first place is because the physical model already created is using it. My first question would be what is the input of a pid controller in modelica? Every input I tried always provides the error message that it is inconsistent 

"[1] 09:58:15 Translation Error
The type of variables engine.flange_b and pid.u
are inconsistent in connect equations.
"

I'm not sure what this even means and how to solve it anymore

Hope this makes better sense now

**Comment by casella** [comment]
@hamadah10 Modelica is not Simulink 😄 

You can't connect a physical port ([engine.flange_b](url)), that carries a pair of coupled variables (angle and torque) with an output signal from a signal processing block such as `pid.u`. You need explicit sensor and actuator models that have one (or more, in some case) physical port, as well as signal ports. In this case, I guess you need to drag a Modelica.Mechanics.Rotational.Sources.Torque model in your model, connect it mechanical port to the engine, and its signal port to the PID output. That is of course if the PID output is meant to be the torque applied to the engine shaft.

Same on the sensor side: if you need to control the engine speed, you need to connect a speed sensor Modelica.Mechanics.Rotational.Sensors.SpeedSensor to the engine shaft, and then its output to the process variable input of the PID controller.

Hope this helps.

**Comment by hamadah10** [comment]
![IMG_8415](https://github.com/OpenModelica/OpenModelica/assets/135056629/e73bbb20-db22-491f-b414-a48720e53e59)

haha. That is true. I am still grasping it and ensuring I use Modelica's 'rules'. Thanks for the help by the way. 

Now, This is how my model looks after taking your comments into consideration! The right side works as intended. As the project I'm working on had it laid out. So, I am currently working on the left side. 




As you can see, I need a relative position/ position sensor and a state estimator called state space in Modelica. Now, I want to connect that portion with the ride-side one. I know that the output of the state space doesn't connect directly to the position sensor. What could be a possible connection to add in order to 'bridge the gap' between them and make it work?  I know that it depends on the problem I am working on, yet, the problem I'm working on only laid out the general guidelines with what I mentioned in my previous comment.

**Comment by casella** [comment]
I'm not sure I understand why you want to convert a signal to a physical model and then back to a signal, but if you want to convert a signal to a linear displacement you can use Modelica.Mechanics.Translational.Sources.Position

**Comment by hamadah10** [comment]
Oh, I never mentioned that probably. This is a physics-based model. Meaning that there is an actual physical model following the same exact sequence and architecture as this. I am trying to model the real physical 'system' to create its digital twin using Modelica. Simply put, I am trying to mimic the actual model. I sat down with the person who created and tested the physical model and told me the steps he took and what he did to make it. Now, I am working on the digital version using Modelica. 

I am trying to follow the real model as closely as possible. Unfortantruly, what he created is not a real car. It is more of a toy car for a different purpose but he used as much knowledge as possible in order to make it accurate to the point where he managed to record the data he wanted from it. 

This is the reason why I wanted to do all this. 

Now, again, I am just following the steps he showed me in his model to recreate a Modelica model of it. Even if I have a slight difference, That is okay as long as the model ends up working. 

I see, I'll try this out and hopefully make it work as intended. I am still not fully certain I understand the full extent of the problem which is causing me all this confusion. 

What I am trying to do in theory is I would connect sensors(control bus?) to a state 'space' or estimator which he called 'flight control' provides coordinates to the PID controller which provides data for a motor then the motor has a feedback loop back to the flight controller for the updated coordinates. 

Thats basically what it should look like. However, this could mean that I would need more sensors or things like the torque to make this whole connector work in Modelica. The person creating this didn't use or consider Modelica or all the factors in this since he meant for this to be more of a toy 'car' after all. 

Hope this makes sense

**Comment by casella** [comment]
Sure it does, but then I don't understand why you want to use physical models to connect the state estimator to the PID controller. There is no physical behaviour there, just signal processing. You can just use blocks à la Simulink for that.

**Comment by hamadah10** [comment]
Let me explain. 

Okay, This is how the creator of the physical model created it. He used a state estimator which he called a flight controller. Simply because he was using GPS coordinates in order to sense the location of the 'vehicle'. He starts with the initial location then he attempts to change that based on the PID controller's feedback loop. 

For me, I am following his footsteps and trying to re-create the model to make it work even if I make changes. Simply because the data is available for everything I mentioned so I could test it. What I am working on could develop into a more refined and complex system (car system, manufacturing plants) but I am learning Modelica because there is a reason for that. I want to, later on, be able to use the code and the features of Modelica for other more complex environments. 

Hope that makes more sense now. Also, I attempted to put the position into the model to connect the state estimator and the position sensor then to the PID controller. However, I got the same error. The model said I still have a translational error between the position and the state estimator...What am I missing? 

I do get that they are fundamentally different in terms of position changing between state estimators back to a linear displacement sounds strange but that's why I'm seeking help as I want to 'fix' this as much as possible.


![currentmodel](https://github.com/OpenModelica/OpenModelica/assets/135056629/c87c4839-e7f9-4e8a-b1eb-9cad129d7fb7)

**Reply by hamadah10** [reply]
@casella 

To make this much easier, I have brought you the system architecture of which the physical model was created. And I'll explain each step.

![the system](https://github.com/OpenModelica/OpenModelica/assets/135056629/359441f9-6fa0-4057-9a6d-a6f6f00c4a3a)


As you can see here, the physical model initially starts with x,y 'gps' coordinates (from the state space) and then provides these coordinates to the PID controller which  'corrects' the path and sends its position and details to the motor and then back to the state estimator as a feedback loop.

I haven't been able to create the feedback loop yet since I'm still struggling with creating the system (the Modelica model)

**Comment by casella** [comment]
I'm sorry but I really do not understand what you want to do 😓. 

As I understand, the part of the system that you surrounded with the purple line is physical, everything else is signal processing. Why are you bringing in physical components also for the other parts of the system, and not just deal with input/output signals as in the block diagram? I really don't get the point of those posistion sensors and actuators. 

In any case, you are trying to connect the physical port of relPositionSensor to a signal input for the ABCD state estimator, which as I wrote above doesn't make sense and is not allowed. You can only connect physical ports to physical ports and signal ports to signal ports.

**Comment by hamadah10** [comment]
i'm sorry about that. 

Okay, the only reason I'm trying to do it this way is simply because this is how the person who created the physical model did it and has the data for it like that. I want to utilize the data later on. 

Now, If we assume that we don't need this and we could just use input and output signals as you mentioned in the block diagram, how can I do that even if I have to give up on some of these sensors and actuators? I am just trying to understand. I know I'm giving you a lot of trouble and confusion and I'm sorry for that. I just want to learn how to do this properly even if as I said it meant that I have to modify this to fit with Modelica's way of functioning as a modeling tool

**Comment by casella** [comment]
I'm sorry again, but I really do not understand your requirements. You have a state estimator called a flight controller that uses GPS coordinates to sense the location of the 'vehicle'. This already doesn't make much sense to me 😅 

Then you represent the 'vehicle' (I guess) in the Modelica model as a 1D rotational mass. I'm totally confused. Isn't the vehicle moving in 3D space? Why is it a 1D thing now? Maybe you should use the 3D Multibody dynamics library for that. And then 
- you put _translational_ (not rotational) 1D sensors in the diagram
- without connecting their physical ports to the physical plant model
- however, you connect the physical ports to the signal input of the state estimator

I'm totally clueless.

Please explain clearly what is the architecture of the system you want to model. Start with a diagram that clearly identifies the physical plant, the sensors, the actuators, the state estimators, and the controller. Then, it's easy to map it into Modelica: physical parts are connected to each other by physical ports, sensors have physical ports to be connected to the physical plant and output signals that are connected to the inputs of signal processing units (state estimators or controllers), actuators have physical ports to be connected to the physical plant, and input signals to be connected to controller outputs.

Modelica.Blocks.Examples.PID_Controller is a nice example of that.

Does this make sense to you?

**Reply by hamadah10** [reply]
Hahha. This is very confusing for everyone. 


Let's start from a conceptual level. 

The system represents an autonomous vehicle. This vehicle has a coordinate that comes from something. To simply know its location. Say its '3D' location. Or if its on a real road, 2D location lets say.

The problem asks to find a way to model this using 1- A State estimator (which is called flight control in the problem) and 2- PID controller. I odn't know why. But that is what was used and the desired system shall have these to be able to feed it data that is available from the test of the actual system that was built using such architecture.

So, 

Again. The system is located somewhere ( the car) how do we know? using its coordinates from the State estimator. Then Initial location we identified from the state estimator should be sensed? or find a way to make the PID control understands that it is in this location and tries to get the cross-track error (the system's distance ) from the desired linear line or destination. We basically want to measure how far it is from the line and want the pid controller to provide the needed control in order to make it drive on a straight line along the desired line we want which is called the path of the vehicle. 

Once we get that, we want to also measure the output we get from the pid controller from its speed and displacement. How fast is the vehicle moving and what its current location? Then we use a motor to basically make that vehicle run. Right? Without It, it won't be able to. 

So, I'm sorry about the long and drawn-out stuff here but I was given the problem with the picture of the system architecture I shared with you. 

IF you were in my place and you receive this picture saying that you need to create a modelica model of it, how would you do it? I gathered all my thoughts about the system and created this 'hopefully' system architecture diagram where i highlighted how exactly things should be and we will go from there. 

The problem statement is basically as follows: We need to create a model to measure the displacement of an autonomous vehicle using two basic things which are the pid controller and state estimator since we already collected this data from a physical device we created.


If I think about it this way, I do would need anything to fulfill that. Forget about sensors and actuators. If the model could work with basic functions, so be it. 

I hope this makes any sense. Haha. I tried to break it down as simple as possible. 

---

Now, forget everything else about the model I created. It is nonsense. Why? Because I am not familiar with Modelica and the problem statement is confusing as you saw in the system architecture that was provided to me.

Even after sitting down with its creator to explain it, I got this: you need to start with a 'sensor' then a flight controller than a PID controller, and then a motor or something then the feedback loop to the state estimator proving updated 'location' to continue that loop and run the tests wanted on it. 


I'm not sure how good is the pid example was as I looked at it and confused me even more XD


![modelica system arc](https://github.com/OpenModelica/OpenModelica/assets/135056629/6f77d0b3-4f13-4c95-9359-a2d6deaaa69f)

I am praying that this makes some sense as I want to make thing more clear

**Comment by casella** [comment]
> I'm not sure how good is the pid example was 

It's as good as it can get. We use it all the time.

> as I looked at it and confused me even more XD

Why? What is not clear in it?

![immagine](https://github.com/OpenModelica/OpenModelica/assets/4233148/db91abbf-62e8-45d9-a29f-9324d8771940)

I'm sorry but this diagram is not really explaining anything. What signals are carried by the arrows? The top one, e.g., says "PID", but that is the name of a block that performs computations, not the name of a signal.

I'm afraid I can't help you much, sorry.
